### ISSUE TITLE
Proposal: Conditional Rendering For Editor Properties Form

### ISSUE THREAD

**Issue by MrWindlike** [webzard-io/sunmao-ui#304]
## Demand Overview
In Some situations, the component lib developers want to make the component property form whether to render depends on another property value. 

Such as the `Table` Component, the `buttonConfig` property should only render when the column's `type` is `button`.

## Scheme Design
Maybe there are three ways to implement this demand:
- Use `Type.Union` to define the properties spec
- Add a new field to properties spec options to decide the property form whether should render
- Let component lib developers use custom widgets to render the properties form

### `Type.Union`
Currently, we already can use `Type.Union` to define property spec to make the form rendering depends on its value.
```js
{
  "type": Type.Union(
    [
      Type.String({
        unionValue: 'text'
      }),
      Type.Object({
        buttonConfig: Type.Object({
          text: Type.String({
            title: 'Text',
          }),
          handlers: Type.Array(EventHandlerSchema),
        }, {
          title: 'Button Config',
          unionValue: 'button'
        }),
      }),
    ]
  ) 
}
```
But this spec would break the type checking because the TypeScript type of `type` in the above example would be: 
```ts
type T = string | { buttonConfig: {...} };
```
So, Maybe this way doesn't work?

### New Field For Conditional Rendering
We can add a new field to spec options to describe when to render this property form. The type definition is:
```ts
interface Condition {
  // the key of the dependent property
  dependsOn: string;
  // rendering form when the dependent property value is equal to this value
  value: string;
}

type UnionConditions = Condition[];
type IntersectionConditions = Condition[];
```
And the example specs are just like these:
```js
// render when: type === 'text' || type === 'button'
{
  "text": Type.String({
    "unionConditions": [{
      "dependsOn": "type",
      "value": "text"
    }, {
      "dependsOn": "type",
      "value": "button"
    }]
  })
}
```
```js
// render when: type === 'text' && enable === true
{
  "text": Type.String({
    "intersectionConditions": [{
      "dependsOn": "type",
      "value": "text"
    }, {
      "dependsOn": "enable",
      "value": true
    }]
  })
}
```
But, how to design the `dependsOn` is a question, we should split it into different situations to talk about it.

#### The Related Properties Are Under The Same Level
When the related properties are under the same level, it would be simple to design the properties spec. The `dependsOn` is set to the key of another property which is under the same level just alright. Such as:
```js
{
  "columns": Type.Array(
  	Type.Object({
      ...,
      "type": Type.KeyOf(...),
      "buttonConfig": Type.Object({...}, {
        "unionConditions": [{
          "dependsOn": "type",
          "value": "button"
        }]
      })
    })
  )
}
```

#### The Related Properties Are Cross The Different Levels
We can set the absolute path to describe which property to depend on. Just like:
```js
{
  "columns": Type.Array(
  	Type.Object({
      ...,
      "type": Type.KeyOf(...),
      "buttonConfig": Type.Object({...}, {
        "unionConditions": [{
          "dependsOn": "columns.type",
          "value": "button"
        }]
      })
    })
  )
}
```
But the above spec has a problem here. You may notice that if a property is `Array` type, it would have the same sub-properties key between different array items, but which property's value should we actually depends on?

I think we should define some rules for the `Array` type properties when depending on it.
- Only allow depending on the property which in the same array item
- Not allow depending on the property under the other `Array` type properties

### Custom Widget
We will support registering the custom widgets for properties form later, and then the component lib developers can freely implement the form widgets that they want. 

However, if the related properties for conditional rendering aren't under the same level, the component lib developers have to implement a large widget to achieve this.

In addition, if a component lib has many properties which only need to render the form depending on different conditions, they have to implement many different widgets for them. That is awful for the component lib, so we better support the common way for the condition rendering.

## Summary
There are three ways that we mention above to implement conditional rendering for properties form, let's take a summary for them.

![image](https://user-images.githubusercontent.com/25414733/155287280-a06070a6-5c0a-42a2-9db6-894567a2a098.png)

In conclusion, I think the second scheme is better.

**Comment by MrWindlike** [comment]
@Yuyz0112 @tanbowensg @xzdry PTAL

**Comment by tanbowensg** [comment]
Very Good! I prefer the second way too.
I think we can implement the "same level" situation first, because the "different level" situation is not as necessary as "same level" and it is totally compatible with "same level".
One more thing, it seems like we only detect whether `dependOn` is equal to `value`. Is there a use case in whitch `dependOn` is a `number` type and the condition is bigger or small than it? Or `dependOn` is `boolean`? Just a thought, not

**Comment by MrWindlike** [comment]
>  it seems like we only detect whether `dependOn` is equal to `value`.

@tanbowensg Yes. Maybe we can support more value comparison operators(such as `gt`, `lt` and etc.) in the future. But I'm not sure `value` or `eq` which name is better for equal comparison.

**Comment by Yuyz0112** [comment]
The second way +1.

Some projects as a reference to the second way:
- https://github.com/CacheControl/json-rules-engine
- https://github.com/RxNT/json-rules-engine-simplified

**Comment by Yuyz0112** [comment]
Oh, looks like the JSON schema also have a conditional semantic built-in: [if-then-else](https://json-schema.org/understanding-json-schema/reference/conditionals.html)

**Comment by Yuyz0112** [comment]
@MrWindlike 

I think we are still going to use the second way. The designs I mentioned are references to the design of the second way. 

We need to check whether the `dependsOn + value` interface is:
- Flexible. Should we consider other logic operators like `OR`? Should we consider other cooperation operators like `IN`, `GREATER_THAN`, `SOME`, `EVERY`?
- Easy to implement, for component developers and the editor runtime.
- Robustness. What will the editor do once a component defined an invalid union condition?

**Comment by MrWindlike** [comment]
> Flexible. Should we consider other logic operators like OR? Should we consider other cooperation operators like IN, GREATER_THAN, SOME, EVERY?

Yes, it's flexible. We can support more operators incrementally in the future and don't worry to make a breaking change. If we want to add more operators, we can add the operators just like this:
```ts
type BaseType = string | number | boolean;
// comparison
type EQ = { 
    // rendering form when the dependent property value is equal to this value
    value: BaseType;
};
type LT = { lt: number };
type GT = { gt: number };
type In = { in: BaseType[] };
type ComparisonOperator = EQ | LT | GT | In;
// logic
type Or = { or: (Logic | ComparisonOperator)[]; }
type And = { and: (Logic | ComparisonOperator)[]; }
type Logic = Or | And;

type Condition = {
  // the key of the dependent property
  dependsOn: string;
} & (ComparisonOperator | Logic);

type UnionConditions = Condition[];
type IntersectionConditions = Condition[];
```

> Easy to implement, for component developers and the editor runtime.

Component developers, only need to follow the design to write JSON just can implement the conditional rendering. In addition, we also export the TypeScript type definition to help the component developers to write the conditional JSON.

The editor needs to implement a function to parse the conditional JSON to real logic execution and another function to get the real value by the path. 

> Robustness. What will the editor do once a component defined an invalid union condition?

Maybe we should return the `false` condition and hide the form widget when the conditional JSONs are invalid?

**Comment by Yuyz0112** [comment]
@MrWindlike Sounds good!

When implementing the condition in the editor, I'm not sure the path like `a.b[0].c` is good or not. The benefit is easy to use and we can use `lodash.get` to read the value, the shortage is it lose some control, and may be hard to reason why a path is invalid.

> Maybe we should return the false condition and hide the form widget when the conditional JSONs are invalid?

Maybe think the condition is invalid and do nothing(do not hide the widgets)?

**Comment by MrWindlike** [comment]
@Yuyz0112 
> may be hard to reason why a path is invalid.

Maybe we can implement a function to read the value and throw the error when the path is invalid instead of using `lodash.get` to solve this?

> Maybe think the condition is invalid and do nothing(do not hide the widgets)?

OK. I think maybe we should support displaying a mask with errors to tell the developers what is wrong with their codes when something is invalid in the develop mode runtime.
